nanopub/publication.py

```diff
@@ -68,32 +94,16 @@ def from_assertion(cls, assertion_rdf, uri=DEFAULT_NANOPUB_URI, introduces_conce
         if attribute_to_profile and profile.get_orcid_id() is not None:
             attributed_to = rdflib.URIRef(profile.get_orcid_id())
 
-        # Make sure passed URI is defrag'd
-        uri = str(uri)
-        uri, _ = urldefrag(uri)
-        this_np = rdflib.Namespace(uri + '#')
-
-        # Replace any blank nodes in the supplied RDF, with a URI derived from the nanopub's uri.
-        # 'Blank nodes' here refers specifically to rdflib.term.BNode objects.
-        # For example, if the nanopub's URI is www.purl.org/ABC123 then the blank node will be replaced with a
-        # concrete URIRef of the form www.purl.org/ABC123#blanknodename where 'blanknodename' is the name of the
-        # the rdflib.term.BNode object. If blanknodename is 'step', then the URI will have a fragment '#step' after it.
-        #
-        # The problem that this is designed to solve is that a user may wish to use the nanopublication to introduce
-        # a new concept. This new concept needs its own URI (it cannot simply be given the nanopublication's URI),
-        # but it should still lie within the space of the nanopub. Furthermore, the URI the nanopub is published
-        # is not known ahead of time. The variable 'this_np', for example, is holding a dummy URI that is swapped
-        # with the true, published URI of the nanopub by the 'np' tool at the moment of publication.
-        #
-        # We wish to replace any blank nodes in the rdf with URIs that are based on this same dummy URI, so that
-        # they too are transformed to the correct URI upon publishing.
-        for s, p, o in assertion_rdf:
-            assertion_rdf.remove((s, p, o))
-            if isinstance(s, rdflib.term.BNode):
-                s = this_np[str(s)]
-            if isinstance(o, rdflib.term.BNode):
-                o = this_np[str(o)]
-            assertion_rdf.add((s, p, o))
+        if introduces_concept and not isinstance(introduces_concept, rdflib.term.BNode):
+            raise ValueError('If you want a nanopublication to introduce a concept, you need to '
+                             'pass it as an rdflib.term.BNode("concept_name"). This will make '
+                             'sure it is referred to from the nanopublication uri namespace upon '
+                             'publishing.')
```

Comment:
I think this makes sense now, but what do you think?

Reply:
I think so. I'm trying to think of any situation in which a nanopub would introduce a concept that is not contained within its assertion subgraph (in some form), in which case you'd obviously have to allow general URIs to be passed. But I don't think that would really ever make sense, so this looks good as is.

I suppose I'd been thinking we could also just accept strings here, but since it would always need to refer to a concept inside the assertion (and since that would need to use ```BNode```) then it is probably best to insist on the user providing a ```BNode``` here. This will save headaches later when they inevitably enter a URI in the string and you get that full URI as a fragment or something horrible like that. I think your error message is clear enough here about what they should do. We will add more to the full documentation once we have it.

Reply:
Nice thoughts. I also thought about passing strings, but indeed then we cannot really check whether it is actually a uri or something else. So good point!